flink-connector-kafka/src/main/java/org/apache/flink/connector/kafka/sink/KafkaRecordSerializationSchemaBuilder.java

```diff
@@ -369,5 +377,23 @@ public ProducerRecord<byte[], byte[]> serialize(
                     value,
                     headerProvider != null ? headerProvider.getHeaders(element) : null);
         }
+
+        @Override
+        public List<LineageDatasetFacet> getDatasetFacets() {
+            List<LineageDatasetFacet> facets = new ArrayList<>();
+            facets.add(new KafkaTopicListFacet(Arrays.asList(topicSelector.apply(null))));
+
+            // gets type information from serialize method signature
+            Arrays.stream(this.valueSerializationSchema.getClass().getMethods())
```

Comment:
Again we should probably check for the serializer to return the TypeInformation directly (by implementing ResultTypeQueryable).
If not we could fallback to extract that as you do, but I'd use things like `org.apache.flink.shaded.guava31.com.google.common.reflect.TypeToken` to be more robust. Your implementation fails if you have some intermediate interface that forward the type parameter.

Reply:
Checking for `ResultTypeQueryable` first is fair. I've switched to guava reflection helpers as suggested, but I am not sure if this helps with `implementation fails if you have some intermediate interface that forward the type parameter.`. Could you provide example of this issue? Not sure if this is now covered or not.